README.md
Pi-hole Admin Dashboard
============
[![Join the chat at https://gitter.im/pi-hole/AdminLTE](https://badges.gitter.im/pi-hole/AdminLTE.svg)](https://gitter.im/pi-hole/AdminLTE?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif "AdminLTE Presentation")](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=3J2L3Z4DHW9UY "Donate")

Using **[AdminLTE](https://almsaeedstudio.com)**, this project will create a Web interface for the ad-blocking Pi-hole: **a black hole for Internet advertisements**.

From this interface, you will be able to see stats on how well your Pi-hole is performing.  You will also be able to update the lists used to block ads.

![Pi-hole Web interface](http://i.imgur.com/5lLAUGo.png)
![Fully responsive](http://i.imgur.com/fHuWR6E.png)

## API
A read-only API can be accessed at `/admin/api.php`. With either no parameters or `api.php?summary` it returns the following JSON:
```JSON
{
	"domains_being_blocked": "136,708",
	"dns_queries_today": "18,108",
	"ads_blocked_today": "14,648",
	"ads_percentage_today": "80.9"
}
```

There are many more parameters, such as `summaryRaw`, `overTimeData`, `topItems`, `recentItems`, `getQueryTypes`, `getForwardDestinations`, `getQuerySources`, and finally `getAllQueries`.

`getAllQueries` can optionally be set with one of these values to return JSON hash ordered as value implies: `orderByClientDomainTime`, `orderByClientTimeDomain`, `orderByTimeClientDomain`, `orderByTimeDomainClient`, `orderByDomainClientTime` or `orderByDomainTimeClient`.


CHANGELOG.MD
#### For the main Pi-hole changelog [see here](https://github.com/pi-hole/pi-hole/wiki/Changelog)

#### [1.5] - 2016-12-XX <-- Work in progress, not yet released!
- `[TWEAK]` - Improve speed of getSummaryData() #223
- `[FIX]` - Secure pause resume plus general security fixes #188
- `[TWEAK]` - Keep main diagram updated #216
- `[TWEAK]` - Add additional pagination control above table in "Query Log" #215
- `[TWEAK]` - Add jsonForceObject flag to enable JSON_FORCE_OBJECT #214
- `[FIX]` - Show the main diagram also if log has been flushed during the day #211
- `[NEW]` - Update ad-serving domains from web UI #201
- `[TWEAK]` - Add option to show *all* recent queries at once #195
- `[NEW]` - Add server-side **password protection for the web interface** #197
- `[NEW]` - Provide links in all three tables on the main page to get the corresponding queries #207
- `[NEW]` - Query list of ad-serving domains via Web UI #209
- `[NEW]` - Help Center to the Web UI #208
- `[NEW]` - App-like behavior on mobile devices #203
- `[NEW]` - Increase temporal resolution in main diagram from 1h to 10min #191
- `[TWEAK]` - Skip last datapoint in main diagram #199
- `[NEW]` - Display CPU and I/O utilization (load) #192
- `[NEW]` - Be able to display CPU temperature in Fahrenheit #196
- `[NEW]` - Add possibility to black-/whitelist entries on the Queries page #194
- `[TWEAK]` - Upgrade Chart.js engine from v1.0.1 to v2.3.0 #187
- `[FIX]` - Fix Gravity Count logic #185
- `[TWEAK]` - Add NewVotex SVG with field ID's #180
- `[FIX]` - Stop PHP from crashing if there aren't enough queries to make a graph #181
- `[FIX]` - Make status logo show in all browsers #179

#### [1.4.4.1] - 2016-10-29
- `[FIX]` - Revert a change that stopped the graphs from working!

#### [1.4.4] - 2016-10-29
- `[NEW]` - A button to pause and resume ad-blocking on your network! [#173](https://github.com/pi-hole/AdminLTE/pull/173) ([@tuplink](https://github.com/tuplink))
- `[FIX]` - Further XSS protection on the whitelist/blacklist page. [#172](https://github.com/pi-hole/AdminLTE/pull/172)


#### [1.4.3.1] - 2016-10-18
- `[FIX]` Revert pull[#154](https://github.com/pi-hole/AdminLTE/pull/154), as it is causing loading issues [#171](https://github.com/pi-hole/AdminLTE/pull/171)

#### [1.4.3] - 2016-10-17
- `[TWEAK]` - No longer show the pop up when update is available, it annoys folk. [#164](https://github.com/pi-hole/AdminLTE/pull/164)
- `[FIX]` - Prevent PHP errors by setting default timezone [#154](https://github.com/pi-hole/AdminLTE/pull/154) (@das7pad)
- `[TWEAK]` - Make arrays backward compatible with PHP 5.3 [#158](https://github.com/pi-hole/AdminLTE/pull/158) (@bcambl)
- `[PERF]` - Performance enhancements in the log parsing [#156](https://github.com/pi-hole/AdminLTE/pull/156) (@akoebbe)
- `[FIX]` - An issue where gravity count was double the actual number [#167](https://github.com/pi-hole/AdminLTE/pull/167)
- `[NEW]` - Add hostname to `Forwarding Destinations` chart if it is found in `/etc/hosts` [#168](https://github.com/pi-hole/AdminLTE/pull/168)
- `[FIX]` - PHP 5.3 support [#160](https://github.com/pi-hole/AdminLTE/pull/160)

#### [1.4.2] - 2016-10-01
- `[FIX]` - Single digit dates not longer cause issues with log parsing... [#151](https://github.com/pi-hole/AdminLTE/pull/151)

#### [1.4.1] - 2016-09-24
- `[FIX]` - Fix temperature output on various devices [#129](https://github.com/pi-hole/AdminLTE/pull/129) (@Fourdee)
- `[TWEAK]` - Remove `pi.hole` and `hostname` entries from web stats for more accurate results. [#134](https://github.com/pi-hole/AdminLTE/pull/134)
- `[TWEAK]` - apple-touch-icon is now png [#139](https://github.com/pi-hole/AdminLTE/pull/139) (@JdenHartog)
- `[FIX]` - Fix for PHP errors in lighttpd error log. [#141](https://github.com/pi-hole/AdminLTE/pull/141)
- `[TWEAK`] - Updated URL and fixed the text styling of GitHub [#135](https://github.com/pi-hole/AdminLTE/pull/135) (@rappo)

#### [1.4] - 2016-08-19
- `[NEW]` - Add CPU temp to status header. The icon becomes red with temps > 45C. [#106](../pull/106)
- `[NEW]` - Log white/blacklist failures. Also allow for a custom hostname to be used to white/blacklist. [#108](../pull/108)
- `[FIX]` - Make tables responsive (query log, top clients, etc). [#111](../pull/111) [#119](../pull/119)
- `[FIX]` - Fix how we handle CORS. Also allow localhost to white/blacklist domains. [#112](../pull/112)
- `[FIX]` - Fix possible XSS attacks through white/blacklist. [#120](../pull/120) [#124](../pull/124)
- `[FIX]` - Fix possible access to unauthorized files through `php/get.php` [#123](../pull/123)
- `[MOD]` - Slight changes to how we get the versions in the footer. [#104](../pull/104)
- `[MOD]` - Change logo to the new one! [#125](../pull/125)

#### [1.3] - 2016-06-11
- `[NEW]` - Whitelist/Blacklist interface! Requires the latest version (currently unreleased) of Pi-hole. [#88](../pull/88)
- `[NEW]` - Added update prompt because of the new functionality [3f66e9](https://github.com/pi-hole/AdminLTE/commit/3f66e9015ab73dadf5cb8f07f3aa3fe30d678fbe)
- `[NEW]` - Implement CSP, move JS out of main HTML, and remove Paypal tracking pixel [#89](../pull/89)
- `[NEW]` - Add update notifications for Pi-hole and the web interface [#92](../pull/92)
- `[FIX]` - Fix XSS bug in hosts list [#84](../pull/84)
- `[FIX]` - Split lines in `/etc/hosts` at any whitespace, not just tabs [#85](../pull/85)
- `[FIX]` - Add `bootstrap.min.css.map` to appease the demands of the dev console [#87](../pull/87)
- `[MOD]` - Add mobile support to the JS disabled alert [#94](../pull/94)

#### [1.2] - 2016-04-16
- `[TWEAK]` - Animations are now turned off, after many requests. Toggle to come in future. [#75](../pull/75)
- `[FIX]` - Status icon actually means something now. Monitors if `dnsmasq` is running. [#71](../pull/71)
- `[FIX]` - Fix for High CPU usage caused by "PHP Warning" flooding the journal[#69](../pull/69)

data.php
<?php
    $log = array();
    $setupVars = parse_ini_file("/etc/pihole/setupVars.conf");
    $hosts = file_exists("/etc/hosts") ? file("/etc/hosts") : array();
    $log = new \SplFileObject('/var/log/pihole.log');

    /*******   Public Members ********/
    function getSummaryData() {
        $domains_being_blocked = gravityCount();

        $dns_queries_today = countDnsQueries();

        $ads_blocked_today = countBlockedQueries();

        $ads_percentage_today = $dns_queries_today > 0 ? ($ads_blocked_today / $dns_queries_today * 100) : 0;

        return array(
            'domains_being_blocked' => $domains_being_blocked,
            'dns_queries_today' => $dns_queries_today,
            'ads_blocked_today' => $ads_blocked_today,
            'ads_percentage_today' => $ads_percentage_today,
        );
    }

    function getOverTimeData() {
        global $log;
        $dns_queries = getDnsQueries($log);
        $ads_blocked = getBlockedQueries($log);

        $domains_over_time = overTime($dns_queries);
        $ads_over_time = overTime($ads_blocked);
        alignTimeArrays($ads_over_time, $domains_over_time);
        return Array(
            'domains_over_time' => $domains_over_time,
            'ads_over_time' => $ads_over_time,
        );
    }

    function getOverTimeData10mins() {
        global $log;
        $dns_queries = getDnsQueries($log);
        $ads_blocked = getBlockedQueries($log);

        $domains_over_time = overTime10mins($dns_queries);
        $ads_over_time = overTime10mins($ads_blocked);
        alignTimeArrays($ads_over_time, $domains_over_time);
        return Array(
            'domains_over_time' => $domains_over_time,
            'ads_over_time' => $ads_over_time,
        );
    }

    function getTopItems() {
        global $log;
        $dns_queries = getDnsQueries($log);
        $ads_blocked = getBlockedQueries($log);

        $topAds = topItems($ads_blocked);
        $topQueries = topItems($dns_queries, $topAds);

        return Array(
            'top_queries' => $topQueries,
            'top_ads' => $topAds,
        );
    }

    function getRecentItems($qty) {
        global $log;
        $dns_queries = getDnsQueries($log);
        return Array(
            'recent_queries' => getRecent($dns_queries, $qty)
        );
    }

    function getIpvType() {
        global $log;
        $dns_queries = getDnsQueries($log);
        $queryTypes = array();

        foreach($dns_queries as $query) {
            $info = trim(explode(": ", $query)[1]);
            $queryType = explode(" ", $info)[0];
            if (isset($queryTypes[$queryType])) {
                $queryTypes[$queryType]++;
            }
            else {
                $queryTypes[$queryType] = 1;
            }
        }

        return $queryTypes;
    }

    function getForwardDestinations() {
        global $log;
        $forwards = getForwards($log);
        $destinations = array();
        foreach ($forwards as $forward) {
            $exploded = explode(" ", trim($forward));
            $dest = hasHostName($exploded[count($exploded) - 1]);
            if (isset($destinations[$dest])) {
                $destinations[$dest]++;
            }
            else {
                $destinations[$dest] = 1;
            }
        }

        return $destinations;

    }

    function getQuerySources() {
        global $log;
        $dns_queries = getDnsQueries($log);
        $sources = array();
        foreach($dns_queries as $query) {
            $exploded = explode(" ", $query);
            $ip = hasHostName(trim($exploded[count($exploded)-1]));
            if (isset($sources[$ip])) {
                $sources[$ip]++;
            }
            else {
                $sources[$ip] = 1;
            }
        }

        global $setupVars;
        if(isset($setupVars["API_EXCLUDE_CLIENTS"]))
        {
            $sources = excludeFromList($sources, "API_EXCLUDE_CLIENTS");
        }

        arsort($sources);
        $sources = array_slice($sources, 0, 10);
        return Array(
            'top_sources' => $sources
        );
    }

    function getAllQueries($orderBy) {
        global $log,$setupVars;
        $allQueries = array("data" => array());
        $dns_queries = getDnsQueriesAll($log);
        $hostname = trim(file_get_contents("/etc/hostname"), "\x00..\x1F");

        if(isset($setupVars["API_QUERY_LOG_SHOW"]))
        {
            if($setupVars["API_QUERY_LOG_SHOW"] === "all")
            {
                $showblocked = true;
                $showpermitted = true;
            }
            elseif($setupVars["API_QUERY_LOG_SHOW"] === "permittedonly")
            {
                $showblocked = false;
                $showpermitted = true;
            }
            elseif($setupVars["API_QUERY_LOG_SHOW"] === "blockedonly")
            {
                $showblocked = true;
                $showpermitted = false;
            }
            elseif($setupVars["API_QUERY_LOG_SHOW"] === "none")
            {
                $showblocked = false;
                $showpermitted = false;
            }
            else
            {
                // Invalid settings, show everything
                $showblocked = true;
                $showpermitted = true;
            }
        }
        else
        {
            $showblocked = true;
            $showpermitted = true;
        }

        foreach ($dns_queries as $query) {
            $time = date_create(substr($query, 0, 16));
            $exploded = explode(" ", trim($query));
            $tmp = $exploded[count($exploded)-4];
            $status = "";

            if (substr($tmp, 0, 5) == "query"){
              $type = substr($exploded[count($exploded)-4], 6, -1);
              $domain = $exploded[count($exploded)-3];
              $client = $exploded[count($exploded)-1];

            }
            elseif ((substr($tmp, 0, 9) == "forwarded" || $exploded[count($exploded)-3] == "pi.hole" || $exploded[count($exploded)-3] == $hostname) && $showpermitted){
              $status="OK";
            }
            elseif ((substr($tmp, strlen($tmp) - 12, 12)  == "gravity.list"  && $exploded[count($exploded)-5] != "read") && $showblocked){
              $status="Pi-holed";
            }

            if ( $status != ""){
              if($orderBy == "orderByClientDomainTime"){
                $allQueries['data'][hasHostName($client)][$domain][$time->format('Y-m-d\TH:i:s')] = $status;
              }elseif ($orderBy == "orderByClientTimeDomain"){
                $allQueries['data'][hasHostName($client)][$time->format('Y-m-d\TH:i:s')][$domain] = $status;
              }elseif ($orderBy == "orderByTimeClientDomain"){
                $allQueries['data'][$time->format('Y-m-d\TH:i:s')][hasHostName($client)][$domain] = $status;
              }elseif ($orderBy == "orderByTimeDomainClient"){
                $allQueries['data'][$time->format('Y-m-d\TH:i:s')][$domain][hasHostName($client)] = $status;
              }elseif ($orderBy == "orderByDomainClientTime"){
                $allQueries['data'][$domain][hasHostName($client)][$time->format('Y-m-d\TH:i:s')] = $status;
              }elseif ($orderBy == "orderByDomainTimeClient"){
                $allQueries['data'][$domain][$time->format('Y-m-d\TH:i:s')][hasHostName($client)] = $status;
              }else{
                array_push($allQueries['data'], array(
                  $time->format('Y-m-d\TH:i:s'),
                  $type,
                  $domain,
                  hasHostName($client),
                  $status,
                  ""
                ));
              }
            }

        }
        return $allQueries;
    }

    /******** Private Members ********/
    function gravityCount() {
        $preEventHorizon = exec("grep -c ^ /etc/pihole/list.preEventHorizon");
        $blacklist = exec("grep -c ^ /etc/pihole/blacklist.txt");
        return ($preEventHorizon + $blacklist);
    }

    function getDnsQueries(\SplFileObject $log) {
        $log->rewind();
        $lines = [];
        foreach ($log as $line) {
            if(strpos($line, ": query[") !== false) {
                $lines[] = $line;
            }
        }
        return $lines;
    }

    function countDnsQueries() {
        return exec("grep -c \": query\\[\" /var/log/pihole.log");
    }

    function getDnsQueriesAll(\SplFileObject $log) {
        $log->rewind();
        $lines = [];
        foreach ($log as $line) {
            if(strpos($line, ": query[") || strpos($line, "gravity.list") || strpos($line, ": forwarded") !== false) {
                $lines[] = $line;
            }
        }
        return $lines;
    }

    function getBlockedQueries(\SplFileObject $log) {
        $log->rewind();
        $lines = [];
        $hostname = trim(file_get_contents("/etc/hostname"), "\x00..\x1F");
        foreach ($log as $line) {
            $line = preg_replace('/ {2,}/', ' ', $line);
            $exploded = explode(" ", $line);
            if(count($exploded) == 8) {
                $tmp = $exploded[count($exploded) - 4];
                $tmp2 = $exploded[count($exploded) - 5];
                $tmp3 = $exploded[count($exploded) - 3];
                //filter out bad names and host file reloads:
                if(substr($tmp, strlen($tmp) - 12, 12) == "gravity.list" && $tmp2 != "read" && $tmp3 != "pi.hole" && $tmp3 != $hostname) {
                    $lines[] = $line;
                };
            }
        }
        return $lines;
    }

    function countBlockedQueries() {
        $hostname = trim(file_get_contents("/etc/hostname"), "\x00..\x1F");
        return exec("grep \"gravity.list\" /var/log/pihole.log | grep -v \"pi.hole\" | grep -v \" read \" | grep -v -c \"".$hostname."\"");
    }

    function getForwards(\SplFileObject $log) {
        $log->rewind();
        $lines = [];
        foreach ($log as $line) {
            if(strpos($line, ": forwarded") !== false) {
                $lines[] = $line;
            }
        }
        return $lines;
    }

    function topItems($queries, $exclude = array(), $qty=10) {
        $splitQueries = array();
        foreach ($queries as $query) {
            $exploded = explode(" ", $query);
            $domain = trim($exploded[count($exploded) - 3]);
            if (!isset($exclude[$domain])) {
                if (isset($splitQueries[$domain])) {
                    $splitQueries[$domain]++;
                }
                else {
                    $splitQueries[$domain] = 1;
                }
            }
        }

        global $setupVars;
        if(isset($setupVars["API_EXCLUDE_DOMAINS"]))
        {
            $splitQueries = excludeFromList($splitQueries, "API_EXCLUDE_DOMAINS");
        }

        arsort($splitQueries);
        return array_slice($splitQueries, 0, $qty);
    }

    function excludeFromList($array,$key)
    {
        global $setupVars;
        $domains = explode(",",$setupVars[$key]);
        foreach ($domains as $domain) {
            if(isset($array[$domain]))
            {
                unset($array[$domain]);
            }
        }
        return $array;
    }

    function overTime($entries) {
        $byTime = array();
        foreach ($entries as $entry) {
            $time = date_create(substr($entry, 0, 16));
            $hour = $time->format('G');

            if (isset($byTime[$hour])) {
                $byTime[$hour]++;
            }
            else {
                $byTime[$hour] = 1;
            }
        }
        return $byTime;
    }

    function overTime10mins($entries) {
        $byTime = array();
        foreach ($entries as $entry) {
            $time = date_create(substr($entry, 0, 16));
            $hour = $time->format('G');
            $minute = $time->format('i');

            // 00:00 - 00:09 -> 0
            // 00:10 - 00:19 -> 1
            // ...
            // 12:00 - 12:10 -> 72
            // ...
            // 15:30 - 15:39 -> 93
            // etc.
            $time = ($minute-$minute%10)/10 + 6*$hour;

            if (isset($byTime[$time])) {
                $byTime[$time]++;
            }
            else {
                $byTime[$time] = 1;
            }
        }
        return $byTime;
    }

    function alignTimeArrays(&$times1, &$times2) {
        if(count($times1) == 0 || count($times2) < 2) {
            return;
        }

        $max = max(array_merge(array_keys($times1), array_keys($times2)));
        $min = min(array_merge(array_keys($times1), array_keys($times2)));

        for ($i = $min; $i <= $max; $i++) {
            if (!isset($times2[$i])) {
                $times2[$i] = 0;
            }
            if (!isset($times1[$i])) {
                $times1[$i] = 0;
            }
        }

        ksort($times1);
        ksort($times2);
    }

    function getRecent($queries, $qty){
        $recent = array();
        foreach (array_slice($queries, -$qty) as $query) {
            $queryArray = array();
            $exploded = explode(" ", $query);
            $time = date_create(substr($query, 0, 16));
            $queryArray['time'] = $time->format('h:i:s a');
            $queryArray['domain'] = trim($exploded[count($exploded) - 3]);
            $queryArray['ip'] = trim($exploded[count($exploded)-1]);
            array_push($recent, $queryArray);

        }
        return array_reverse($recent);
    }

    function hasHostName($var){
        global $hosts;
        foreach ($hosts as $host){
            $x = preg_split('/\s+/', $host);
            if ( $var == $x[0] ){
                $var = $x[1] . "($var)";
            }
        }
        return $var;
    }
?>
